### PR TITLE
fictitious should be an Identifiable's attribute, not only a Switch's attribute

### LINKED ISSUE #1104: fictitious should be an Identifiable's attribute, not only a Switch's attribute
* **Do you want to request a *feature* or report a *bug*?**
Feature

* **What is the current behavior?**
Only `Switch` and `Load` can be fictitious.

* **What is the expected behavior?**
Basically any equipment (`ShuntCompensator`, `Load`, `Generator`, `Switch`...) can be fictitious... This attribute should probably belong to `Identifiable`.

* **What is the motivation / use case for changing the behavior?**
This information can result in a different display in visualization tools (fictitious equipment in a different color for example).

* **Please tell us about your environment:**
  - PowSyBl Version: 3.2.0-SNAPSHOT

### PR DESCRIPTION
**Pull request by tadam50**
Signed-off-by: Thomas ADAM <tadam@silicom.fr>

**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem ?** *If so, link to this issue using `'#XXX'` and skip the rest*
fix #1104

**Does this PR introduce a breaking change or deprecate an API?** *If yes, check the following:*
- [ ] The *Breaking Change* or *Deprecated* label has been added
- [ ] The migration guide has been updated in the github wiki *(What changes might users need to make in their application due to this PR?)*

**Other information**:

(if any of the questions/checkboxes don't apply, please delete them entirely)

### CODE REVIEW

File: iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/AbstractIdentifiable.java [hunk 1]
```diff
@@ -39,6 +46,18 @@ public String getName() {
...
+    @Override
+    public boolean isFictitious() {
+        return fictitious;
+    }
+
+    @Override
+    public void setFictitious(boolean fictitious) {
+        boolean oldValue = this.fictitious;
+        this.fictitious = fictitious;
+        getNetwork().getListeners().notifyUpdate(this, "fictitious", oldValue, fictitious);
```
Comment (mathbagu) [inline iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/AbstractIdentifiable.java]: You should compare old and new value and notify only if there is a modification? Check how it’s done for other attributes

File: iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/TieLineImpl.java [hunk 2]
```diff
@@ -6,11 +6,11 @@
  */
 package com.powsybl.iidm.network.impl;
 
-import java.util.Objects;
-
 import com.powsybl.iidm.network.TieLine;
 import com.powsybl.iidm.network.ValidationException;
 
+import java.util.Objects;
+
```
Comment (mathbagu) [inline iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/TieLineImpl.java]: You should revert this change

File: iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/TwoWindingsTransformerImpl.java [hunk 3]
```diff
@@ -6,15 +6,11 @@
...
+import com.powsybl.iidm.network.*;
+
 import java.util.HashSet;
 import java.util.Set;
 
-import com.powsybl.iidm.network.ConnectableType;
-import com.powsybl.iidm.network.Identifiable;
-import com.powsybl.iidm.network.TapChanger;
-import com.powsybl.iidm.network.TwoWindingsTransformer;
-import com.powsybl.iidm.network.ValidationUtil;
```
Comment (mathbagu) [inline iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/TwoWindingsTransformerImpl.java]: Maybe there is something in your IDE configuration. The import order should remain the same. Maybe we can share our configuration

File: iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/HvdcLineAdderAdapter.java [hunk 4]
```diff
@@ -130,6 +133,12 @@ public HvdcLineAdder setEnsureIdUnicity(final boolean ensureIdUnicity) {
         return this;
     }
 
+    @Override
+    public HvdcLineAdder setFictitious(final boolean fictitious) {
+        this.fictitious = fictitious;
+        return this;
+    }
```
Comment (mathbagu) [inline iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/HvdcLineAdderAdapter.java]: Do you confirm that HvdcLineAdderAdapter doesn’t inherit from a class which is responsible of the fictitious attribute

Comment (tadam50) [reply iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/HvdcLineAdderAdapter.java]: `HvdcLineAdderAdapter` does not inherit from `AbstractIdentifiableAdderAdapter`
To be discussed:
I can create a new class which is responsible of the id, name, ensureIdUnicity, fictitious & MergingViewIndex
`HvdcLineAdderAdapter` & `LineAdderAdapter` could inherit from it

File: iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/AbstractIdentifiable.java (hunk 1 above)
Comment (tadam50) [reply iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/AbstractIdentifiable.java]: `NetworkListenerList::notifyUpdate` already check if there is a modification

File: iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/TieLineImpl.java (hunk 2 above)
Comment (tadam50) [reply iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/TieLineImpl.java]: Sorry, my bad 
Use Intellij configuration "Optimize imports on the fly"

File: iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/TwoWindingsTransformerImpl.java (hunk 3 above)
Comment (tadam50) [reply iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/TwoWindingsTransformerImpl.java]: Sorry, my bad
Use Intellij configuration "Optimize imports on the fly"

**Comment by tadam50** [comment]
> Do you miss to update the XML export/import?

Good catch !

File: iidm/iidm-api/src/main/java/com/powsybl/iidm/network/IdentifiableAdder.java [hunk 5]
```diff
@@ -18,4 +18,5 @@
 
     T setName(String name);
 
+    T setFictitious(boolean fictitious);
```
[the remark this answers is not in the document]
Comment (tadam50) [reply iidm/iidm-api/src/main/java/com/powsybl/iidm/network/IdentifiableAdder.java]: Good catch

File: iidm/iidm-api/src/main/java/com/powsybl/iidm/network/Identifiable.java [hunk 6]
```diff
@@ -74,4 +74,18 @@ default Properties getProperties() {
      * Get properties key values.
      */
     Set<String> getPropertyNames();
+
+    /**
+     * Get the fictitious status of the switch
+     */
+    default boolean isFictitious() {
+        throw new UnsupportedOperationException();
+    }
```
Comment (mathbagu) [inline iidm/iidm-api/src/main/java/com/powsybl/iidm/network/Identifiable.java]: This method could return false by default instead of throwing an exception.
@MioRtia What do you think?

File: iidm/iidm-api/src/main/java/com/powsybl/iidm/network/Identifiable.java [hunk 7]
```diff
@@ -74,4 +74,18 @@ default Properties getProperties() {
      * Get properties key values.
      */
     Set<String> getPropertyNames();
+
+    /**
+     * Get the fictitious status of the switch
```
Comment (mathbagu) [inline iidm/iidm-api/src/main/java/com/powsybl/iidm/network/Identifiable.java]: Fix the javadoc

File: iidm/iidm-api/src/main/java/com/powsybl/iidm/network/Identifiable.java [hunk 8]
```diff
@@ -74,4 +74,18 @@ default Properties getProperties() {
...
+     */
+    default boolean isFictitious() {
+        throw new UnsupportedOperationException();
+    }
+
+    /**
+     * Set the fictitious status of the switch
```
Comment (mathbagu) [inline iidm/iidm-api/src/main/java/com/powsybl/iidm/network/Identifiable.java]: Fix the javadoc

File: iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/MergedLine.java [hunk 9]
```diff
@@ -432,6 +432,21 @@ public boolean hasProperty(final String key) {
...
+
+    @Override
+    public void setFictitious(boolean fictitious) {
+        if (dl1.isFictitious() != fictitious) {
+            dl1.setFictitious(fictitious);
+        }
+        if (dl2.isFictitious() != fictitious) {
+            dl2.setFictitious(fictitious);
+        }
+    }
```
Comment (mathbagu) [inline iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/MergedLine.java]: I think the tests are useless: why don't you delegate the call automatically

**Comment by mathbagu** [comment]
@MioRtia: I requested the change of the Javadoc. My first comment about the default behavior of getFictitious is not blocking. Same for the last one.

File: iidm/iidm-api/src/main/java/com/powsybl/iidm/network/Identifiable.java (hunk 6 above)
Comment (MioRtia) [reply iidm/iidm-api/src/main/java/com/powsybl/iidm/network/Identifiable.java]: Both solutions are good to me!

**Comment by MioRtia** [comment]
> @MioRtia: I requested the change of the Javadoc. My first comment about the default behavior of getFictitious is not blocking. Same for the last one.

Okay for me, thank you @tadam50 for your quick corrections, I am merging this PR.
